ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitterSpec.scala

```diff
@@ -66,6 +69,77 @@ class TransactionCommitterSpec extends WordSpec with Matchers with MockitoSugar
       Conversions.configurationStateKey -> Some(configurationStateValue),
       dedupKey -> Some(dedupValue))
+  private val aRichTransactionTreeSummary = {
+    val roots = Seq("Exercise-1", "Fetch-1", "LookupByKey-1", "Create-1")
+    val nodes: Seq[TransactionOuterClass.Node] = Seq(
+      createNode("Fetch-1")(_.setFetch(fetchNodeBuilder)),
+      createNode("LookupByKey-1")(_.setLookupByKey(lookupByKeyNodeBuilder)),
+      createNode("Create-1")(_.setCreate(createNodeBuilder)),
+      createNode("LookupByKey-2")(_.setLookupByKey(lookupByKeyNodeBuilder)),
+      createNode("Fetch-2")(_.setFetch(fetchNodeBuilder)),
+      createNode("Create-2")(_.setCreate(createNodeBuilder)),
+      createNode("Fetch-3")(_.setFetch(fetchNodeBuilder)),
+      createNode("Create-3")(_.setCreate(createNodeBuilder)),
+      createNode("LookupByKey-3")(_.setLookupByKey(lookupByKeyNodeBuilder)),
+      createNode("Exercise-2")(
+        _.setExercise(
+          exerciseNodeBuilder.addAllChildren(
+            Seq("Fetch-3", "Create-3", "LookupByKey-3").asJava
+          ))),
+      createNode("Exercise-1")(
+        _.setExercise(
+          exerciseNodeBuilder.addAllChildren(
+            Seq("LookupByKey-2", "Fetch-2", "Create-2", "Exercise-2").asJava
+          )))
+    )
+    val tx = TransactionOuterClass.Transaction
+      .newBuilder()
+      .addAllRoots(roots.asJava)
+      .addAllNodes(nodes.asJava)
+      .build()
+    val outTx = aDamlTransactionEntry.toBuilder.setTransaction(tx).build()
+    DamlTransactionEntrySummary(outTx)
+  }
+
+  private def createNode(nodeId: String)(
+      nodeImpl: TransactionOuterClass.Node.Builder => TransactionOuterClass.Node.Builder) =
+    nodeImpl(TransactionOuterClass.Node.newBuilder().setNodeId(nodeId)).build()
+
+  private def fetchNodeBuilder = TransactionOuterClass.NodeFetch.newBuilder()
+
+  private def exerciseNodeBuilder = TransactionOuterClass.NodeExercise.newBuilder()
+
+  private def createNodeBuilder = TransactionOuterClass.NodeCreate.newBuilder()
+
+  private def lookupByKeyNodeBuilder = TransactionOuterClass.NodeLookupByKey.newBuilder()
+
+  "removeUnnnecessaryNodes" should {
+    "remove `Fetch` and `LookupByKey` nodes from transaction tree" in {
+      val context = new FakeCommitContext(recordTime = None)
+
+      instance.removeUnnecessaryNodes(context, aRichTransactionTreeSummary) match {
+        case StepContinue(_) => fail("should be StepStop")
+        case StepStop(logEntry) =>
+          val transaction = logEntry.getTransactionEntry.getTransaction
+          transaction.getRootsList.asScala should contain theSameElementsInOrderAs Seq(
+            "Exercise-1",
+            "Create-1")
+          val nodes = transaction.getNodesList.asScala
+          nodes.map(_.getNodeId) should contain theSameElementsInOrderAs Seq(
+            "Create-1",
+            "Create-2",
+            "Create-3",
+            "Exercise-2",
+            "Exercise-1")
+          nodes.head.hasCreate shouldBe true
```

Comment:
I don't think this assertion is necessary, i.e., the node ID already implies the type by construction.

Reply:
Removed.